src/main/java/picard/analysis/FastWgsMetricsCollector.java

```diff
@@ -0,0 +1,195 @@
+        long processedLoci = counter;
+        readsNames.put(record.getReadName(), recordsAndOffsetsForName);
+        final byte[] qualities = record.getBaseQualities();
+        final byte[] bases = record.getRecord().getReadBases();
+        for (int i = 0; i < record.getLength(); i++) {
+            final int index = i + position;
+            if (isReferenceBaseN(index, ref)) {
+                continue;
+            }
+            final byte quality = qualities[i + record.getOffset()];
+            if (quality <= 2) {
+                basesExcludedByBaseq++;
+            } else {
+                if (unfilteredDepthSize.get(index) < coverageCap) {
+                    unfilteredBaseQHistogramArray[quality]++;
+                    unfilteredDepthSize.increment(index);
+                }
+                if (quality < collectWgsMetrics.MINIMUM_BASE_QUALITY || SequenceUtil.isNoCall(bases[i + record.getOffset()])){
+                    basesExcludedByBaseq++;
+                } else {
+                    final int bsq = excludeByQuality(recordsAndOffsetsForName, index);
+                    if (recordsAndOffsetsForName.size() - bsq > 0) {
+                        basesExcludedByOverlap++;
+                    } else {
+                        pileupSize.increment(index);
+                    }
+                }
+            }
+            if (isTimeToStop(++processedLoci)) {
+                break;
+            }
+        }
+        recordsAndOffsetsForName.add(record);
+    }
+
+    private void removeRecordFromMap(EdgingRecordAndOffset record, Set<EdgingRecordAndOffset> recordsAndOffsetsForName) {
+        if (recordsAndOffsetsForName.size() == 1) {
+            readsNames.remove(record.getReadName());
+        } else {
+            recordsAndOffsetsForName.remove(record.getStart());
+        }
+    }
+
+    /**
+     * Prepares the accumulator objects to process a new {@link htsjdk.samtools.util.AbstractLocusInfo}.
+     * If we switch to a new sequence, all accumulators are cleared.
+     *
+     * @param info the next {@link htsjdk.samtools.util.AbstractLocusInfo} to process
+     */
+    private void prepareCollector(AbstractLocusInfo<EdgingRecordAndOffset> info) {
+        if (readsNames == null) {
+            readsNames = new HashMap<>();
+        }
+        if (previousSequenceIndex != info.getSequenceIndex()) {
+            readsNames.clear();
+            counterManager.clear();
+            previousSequenceIndex = info.getSequenceIndex();
+        }
+        counterManager.checkOutOfBounds(info.getPosition());
+    }
+
+    private int excludeByQuality(final Set<EdgingRecordAndOffset> setForName, int position) {
+        int bsq = 0;
+        for (EdgingRecordAndOffset recordAndOffset : setForName) {
+            if (position - recordAndOffset.getRefPos() >= recordAndOffset.getLength()
```

Comment:
first part of if statement isn't clear here. can you add a comment explaining it?